### ISSUE TITLE
Add Contact Email Field To A Publisher Account

### ISSUE THREAD

**Issue by iatisupport** [okfn/ckanext-iati#71]
A single contact email address for a publisher needs to be displayed under the organisation website on the right hand sidebar when viewing publisher details  eg http://iatiregistry.org/publisher/dipr

To do this we propose that the publisher admin 'Contact' field is renamed address and an additional new field called 'Contact Email' is added underneath this 'Address' (formally Contact) field. Also the 'Contact Email' field should be mandatory.

**Comment by dalepotter** [comment]
This issue appears to be have been addressed in cbd2663 and f76abac. 

Once deployed to the live instance, this issue can be closed

**Comment by chrisviderum** [comment]
@dalepotter @wendyrogers 

This field should be now available.

**Comment by wendyrogers** [comment]
Thank you @chrisviderum . I can see the new field which is good but it should be mandatory when the publisher account is created or updated. I can currently update an existing publisher account with an empty field?

**Comment by chrisviderum** [comment]
@wendyrogers then we will make it mandatory. I will update this ticket as soon as this was done.

**Comment by chrisviderum** [comment]
@wendyrogers field should be mandatory now.

iati-staging.ckan.io

Does it work like you would like to have it now?

**Comment by wendyrogers** [comment]
Thanks @chrisviderum and I have retested. However, it appears that I can actually enter a non valid email address in the field. Please can the structure of the input be tested for a valid email address?

Also, will this change have any impact on any API call or has it only been currently implemented thorugh the web user interface?

**Comment by chrisviderum** [comment]
@wendyrogers having a validation process for the field wasn't part of the development scope until now and wasn't considered for the estimation given for this work.

I will ask my development how long it takes to add a feature like that. Of course, I will ask them about the API question, too.

**Comment by chrisviderum** [comment]
It will take us 1 hour to add this functionality. Is it fine for you, if we add it?

There won't be an impact on the API by this.

**Comment by wendyrogers** [comment]
Many thanks @chrisviderum and yes the 1 hour estimate is fine for adding the validation

**Comment by chrisviderum** [comment]
@wendyrogers validation should be implemented.

**Comment by wendyrogers** [comment]
thanks @chrisviderum . I have now tested and validation is working but could the error message be something more specific and user friendly. So rather than 'Please match the required format' something like ' Please provide a valid email address. The email address should be for a mailbox that is regularly monitored'

**Comment by chrisviderum** [comment]
@wendyrogers we have made this change as you requested. Can we close this one then?

Then we are able to push 6 improvements to the registry!

**Comment by wendyrogers** [comment]
@chrisviderum I have now retested. However I am now getting the required error message for valid email addresses such as enquiries@beesabroad.org.uk and wendy.r@devinit.org?

**Comment by chrisviderum** [comment]
Dear Wendy,

I tested it with some different email addresses and it worked fine. We will try it out and give you feedback soon.

Best,

Chris

**Comment by chrisviderum** [comment]
@wendyrogers have you made you tests on iati-staging.ckan.io ?

I have tested it there for the Abt Associates publisher: https://iati-staging.ckan.io/publisher/abt

tried both email addresses and it worked. Can you give me a step by step what you are doing?

Thanks and best,

Chris

**Comment by wendyrogers** [comment]
Thanks @chrisviderum . While logged in as System Admin user I am still getting an error message when trying to update email address and clicking 'Update Publisher'. See screen print 
![image](https://cloud.githubusercontent.com/assets/5500787/20559740/e1518a90-b16d-11e6-9b95-b6a829f763be.png)

**Comment by pmartignoli** [comment]
Hi @wendyrogers ,
Would you please be able to check this? I have been able to add email addresses without receiving error messages.
Thanks very much!
Paul
Viderum Customer Service

**Comment by dalepotter** [comment]
@PetyaKangalova would you be able to take a look to check this on the staging site: https://iati-staging.ckan.io/

**Comment by PetyaKangalova** [comment]
@chrisviderum @pmartignoli @dalepotter  I just checked on the staging site and unfortunately the majority of publisher pages come up with "internal server" error. It seems to me that this relates to the "internal server error" issue #111 on  Github. Publishers that haven't added **Contact email** are not showing on the staging site.

See example screenshots below- for the same publisher I get a server error on the staging site but can access the page on the live site:
![image](https://cloud.githubusercontent.com/assets/24607542/24545696/12be06ac-1600-11e7-8c6f-cd0bc164b880.png)
![image](https://cloud.githubusercontent.com/assets/24607542/24545718/302002b8-1600-11e7-8251-d6325d5fc167.png)

Once Github issue #111 has been resolved and I am able to access the publisher pages on the staging website, I will be happy to test again and update you if the error occurs again.

**Comment by PetyaKangalova** [comment]
@pmartignoli  I just checked again a few random publishers on the staging site and Error 500- Internal server error still remains. Could you please look into this? See an example below. I works fine on the live site but the staging site still gives an error.
Link: https://iati-staging.ckan.io/publisher/ai_1064413

![image](https://cloud.githubusercontent.com/assets/24607542/26158237/22d332f0-3b13-11e7-9f36-ebc8b1142f57.png)

**Comment by pmartignoli** [comment]
@PetyaKangalova, thanks for this. We will migrate the production database to staging next week, which should be beneficial for a number of reasons, including hopefully solving this issue. We will revisit this once the migration has been made.

**Comment by PetyaKangalova** [comment]
@pmartignoli  Many thanks. Let me know once done and I will be happy to test it again.

**Comment by pmartignoli** [comment]
@PetyaKangalova  @dalepotter 
I have emailed an Excel sheet containing a list of 443 publishers with no contact email address. It has the name and id of the publisher, whether they've published a dataset or not.

If you are able to fill in the publisher_contact_email column, we can handle the updates from there.

This will ensure there will be no further server errors due to missing emails (finally close 111) and we can remove the workaround and close 71 as all current publishers will have an email and no future publisher will be able to register without one.

If the emails are not available, a placeholder can be added however the correct email would be preferable.

I look forward to hearing from you,

Paul

**Comment by pmartignoli** [comment]
@PetyaKangalova.
Following on from our call last Friday regarding placeholders being added for publishers with missing email fields.

In October 2016, the following pull-request was made:

https://github.com/ViderumGlobal/ckanext-iati/pull/97/commits/422ac73ed35710477de0859bb42dc4fd8d59c687

This pull request adds placeholder text ('Email not found') to the publisher email field that users see when the page is rendered if the publisher contact email is missing so that the page can be rendered successfully. It does not add a placeholder email address (like pubcontactmissing@iatiregistry.org) to the database.

Could this be the change regarding adding placeholders you mentioned? I believe we also discussed adding placeholders to all missing fields as a possibility earlier this year, however had not decided to make it an action point. Please correct me if I am wrong. 

Regards,
Paul

**Comment by PetyaKangalova** [comment]
@pmartignoli  Thanks for the clarification.  Yes, what we want to is to have a placeholder text for **all** publisher accounts where there is **no** Contact Email for publisher (see below).


![image](https://user-images.githubusercontent.com/24607542/29464282-5bfd1f2e-842d-11e7-96aa-216a0192ff7d.png)
What I don't understand is why there are still publishers with no placeholder text if the pull-request specified above ([422ac73](https://github.com/ViderumGlobal/ckanext-iati/commit/422ac73ed35710477de0859bb42dc4fd8d59c687)) was implemented. Has it been updated since October 2016?

Just to clarify again the reason for adding a placeholder text is due to the fact that the Contact Email was made mandatory. As a result, when publishers with NO Contact Emails were updating their data, the IATI Registry was showing an internal server error on their page.  As such, could we now make sure that all publishers with no Contact Email have a placeholder text added? 

What you are referring as 

> email address in the database

I presume is the email publishers add to individual datasets. It will  be difficult to identify which email addresses can be linked to the publisher accounts and we would not want to do that without the publisher's consent, so I suggest that we go ahead with the placeholder text on all publisher accounts, unless @dalepotter has other views.

I will not be able to attend our call next week but as this is on the low priority list I will be happy to chat again in early September when I am back from leave.

Thanks,
Petya

**Comment by JGulic** [comment]
@PetyaKangalova @dalepotter we've created a script that goes through all the publisher's and inserts a 'Email Not Found' placeholder if the publisher doesn't have an email address. We've run it successfully on staging and you can check out the publishers there.
If everything is ok for you with this, we can move it to production.

**Comment by PetyaKangalova** [comment]
@JGulic Could you please refer me to the repository where you have created the script? I can only identify an old one in [here](https://github.com/ViderumGlobal/ckanext-iati/commit/422ac73ed35710477de0859bb42dc4fd8d59c687) and I am sure there is a newer one who have written.

Could you also let me know if you have done some testing on your side for this specific script? I have done one or two manual checks on the staging site, which seem to be working properly, but manual checks are not reliable as there are quite a lot of publishers with missing contact emails.

**Comment by JGulic** [comment]
@PetyaKangalova the new script is [here](https://github.com/ViderumGlobal/ckanext-iati/blob/publishers-email-script/scripts/fill_empty_publisher_emails.py), not merged to master yet.

As I mentioned before, the script goes through every publisher in the registry and checks if the publisher has an email address. If there isn't an email address the script will insert 'Email Not Found' in the email field. This way we can be sure that if a publisher tries to manage their data, they will have to enter an email address in order to make the changes they want, because the placeholder will trigger the email validation when the user will try to save the data.

We've tested out the script locally with the latest IATI database, as well as on the staging portal and we didn't encounter any problems.
Currently there are 439 publishers in the registry that don't have an email address. You can find a full list of them [here](https://gist.github.com/JGulic/2c379a949c56f9a6de8c534754f2a461).

**Comment by andylolz** [comment]
I ran the script and found 499 publishers with no email address.

BUT I noticed that the `publisher_contact` field frequently includes an email address. Is this of interest / use? Here’s a script (with output) that additionally makes a naive attempt to regex this information from the `publisher_contact` field:
https://gist.github.com/andylolz/8f7df5df777cef95b4e1638aa7dbf7ae

In this case, only 212 email addresses are missing.

**Comment by PetyaKangalova** [comment]
@andylolz  Many thanks for the suggestion and running the script. We actually considered the same option before, however, as the email has been added in their data, we cannot assume that this is the email they would like to add to their publisher account on the Registry (which I know is probably the case), or have the permission to do so. What I could do is send an email to all the identified publisher contacts and ask them to update their Registry account, which will result in at least a few organisation updating their profile. 

@JGulic Thanks for sharing the script. Yes, could you please go ahead and merge to master? Once that is done successfully, could you please add a comment to this post with a link and close the issue?

Thanks again!

**Comment by andylolz** [comment]
> What I could do is send an email to all the identified publisher contacts and ask them to update their Registry account, which will result in at least a few organisation updating their profile.

I’d suggest making it a passive change i.e. “we’ve made this change; please correct it if it’s wrong.” That’s likely to be more convenient for publishers, and will almost certainly result in higher conversion.

> We actually considered the same option before, however, as the email has been added in their data, we cannot assume that this is the email they would like to add to their publisher account on the Registry

The script I used doesn’t look at the `contact-info` field in the IATI data. Is that what you mean? If so, I agree that that field shouldn’t be used for this, because it’s potentially unrelated.

My script looks at two fields in the registry metadata for the publisher. [Take AidData, for example](https://iatiregistry.org/publisher/about/aiddata):
![screen shot 2017-11-14 at 13 51 30](https://user-images.githubusercontent.com/464193/32783255-ff545700-c942-11e7-96b9-0edf1f20ab28.png)

The left side of the screen provides a contact email address; the right side of the screen says “Email not found”. My suggestion is: when there’s something in the field on the left that looks like an email address, copy it to the field on the right. [In my opinion, when this new ‘Contact Email’ field was added, something similar to this should have been done. I acknowledge you would have to do it carefully, but it’s still eminently doable.]

I’d actually go even further, and check contact email addresses in the dataset metadata. If those email addresses are the same for all datasets from a publisher, I think it would be reasonable to assume that’s the best contact for the publisher.

**Comment by PetyaKangalova** [comment]
@andylolz  Many thanks for the clarification. 

> The script I used doesn’t look at the contact-info field in the IATI data. 

I indeed thought that the script takes the email from the XML file, rather than the metadata.

Using the **Contact** field in the metadata (your existing script) is a very good idea and I think we should proceed with adding the emails you have identified. I can then send an email to inform them that we have updated their email and ask them to correct if wrong.

Then, on using the email from the dataset metadata (showing an example below for Viderum's reference), I would also support that. Would @andylolz  or  @JGulic be able to run a script to identify those emails for publishers where the Publisher email is missing? If possible, that will be great.


![image](https://user-images.githubusercontent.com/24607542/32940878-e0eafb8a-cb7b-11e7-8a37-ce1dfa0ea7f5.png)


@JGulic  I think once we have both scripts with emails (one taking the emails from Andy's script) and hopefully another one where ( for those publishers that we still don't have emails) we can identify an email in the dataset metadata (example above), then we are good to go. @JGulic  Let me know if that makes sense or if you have any questions.

**Comment by JGulic** [comment]
@PetyaKangalova there are few problems that I can see with using the contact field in the metadata:
- One file can have more then one contact-info element in the XML file, which could contain different email addresses and in this case we would not know which one to get
- One Publisher can have many datasets with different resources, and all of them may contain a different information in the contact-info element.
So I don't think it's a good idea to use the contact-info element of the metadata.

@andylolz can you share the code from the scrip that you were running to check the publishers, so we can take a look?

**Comment by andylolz** [comment]
> @andylolz can you share the code from the scrip that you were running to check the publishers, so we can take a look?

It’s included in the gist I shared:
https://gist.github.com/andylolz/8f7df5df777cef95b4e1638aa7dbf7ae#file-script-py

**Comment by JGulic** [comment]
@andylolz thank you, sorry for missing it

**Comment by andylolz** [comment]
>  * One file can have more then one contact-info element in the XML file, which could contain different email addresses and in this case we would not know which one to get

I think you’re talking about the `contact-info` element in the data, rather than the metadata. Petya and I both agreed we don’t want to use the `contact-info` element – instead, we’re talking about the metadata here. It’s probably easiest to understand from the two images (of the AidData publisher page and ZOA activity file dataset page) in the comments above.

>  * One Publisher can have many datasets with different resources, and all of them may contain a different information in the contact-info element.
> So I don't think it's a good idea to use the contact-info element of the metadata.

I explained above one way of dealing with this case (emphasis added):

> I’d actually go even further, and check contact email addresses in the dataset metadata. *If those email addresses are the same for all datasets from a publisher*, I think it would be reasonable to assume that’s the best contact for the publisher.

**Comment by andylolz** [comment]
@PetyaKangalova wrote:
> Then, on using the email from the dataset metadata (showing an example below for Viderum's reference), I would also support that. Would @andylolz or @JGulic be able to run a script to identify those emails for publishers where the Publisher email is missing? If possible, that will be great.

Okay – here’s an improved version of the script:
https://gist.github.com/andylolz/41be8ef286a6c5dd2eafd8f04817a186#file-script2-py

This one looks at datasets, too. The output is in the same gist.

**Comment by andylolz** [comment]
I’m just looking at the data use task force objectives – I see objective 2 involves improving data quality “through the development of feedback mechanisms”.

A means of contacting a publisher to flag data quality issues seems like a prerequisite for this. This address could be used by d-portal (and other portals) to populate the contact address for the bar at the bottom (rather than using support@iatistandard.org)
![screen shot 2017-11-24 at 17 32 37](https://user-images.githubusercontent.com/464193/33220043-8e071800-d13d-11e7-9820-7c31e15c906e.png)

**Comment by PetyaKangalova** [comment]
@andylolz Many thanks for updating the script so that it also checks for emails in the dataset metadata. @JGulic Could you please check and run the script? We are not waiting for anything else now- I am happy to use the emails that Andy has identified as they are specified by publishers in their metadata that is publically available. Please, could you also update this Github issue once done?

@andylolz On your[ latest comment,](https://github.com/ViderumGlobal/ckanext-iati/issues/71#issuecomment-346874346) in relation to d-portal, could you please add to the d-portal Github repository so that it can be addressed there and we can close this issue with Viderum here (once the emails have been updated)?  You are right that the feedback mechanism can be improved and I have a suggestion on how we can change the text on d-protal

**Comment by goranmaxim** [comment]
I see the following columns produced by @andylolz's script:
org_name	original	publisher_contact	unique_dataset_email	dataset_emails

https://gist.github.com/andylolz/41be8ef286a6c5dd2eafd8f04817a186

Are we populating the publisher_contact field (where empty) with a value from "original", "unique_dataset_email" or "dataset_emails" (in that order)?

**Comment by PetyaKangalova** [comment]
@goranmaxim  Apologies for the significant delay in responding. The publisher_contact field should be populated with a value from "publisher_contact" or  "unique_dataset_email" (in that order).

The original is the email for organisation that already have an email and dataset_emails has been used to identify if there is a unique_dataset_email. Let me know if there are any other questions.

**Comment by andylolz** [comment]
> in relation to d-portal, could you please add to the d-portal Github repository so that it can be addressed there and we can close this issue with Viderum here (once the emails have been updated)?

@PetyaKangalova: Absolutely – Once the emails have been updated and this issue is closed, I’ll add the issue to d-portal. Otherwise it’s just blocked by this issue anyway.

**Comment by goranmaxim** [comment]
The field has been updated for all publishers as per 
https://gist.github.com/andylolz/41be8ef286a6c5dd2eafd8f04817a186
with either "publisher_contact" or "unique_dataset_email".
The only ones that have an empty publisher contact field are the ones we didn't have data to populate it from.
